src/kernel/pci.zig

```diff
@@ -0,0 +1,310 @@
+    RevisionId = 0x08,
+    ProgrammingInterface = 0x09,
+    Subclass = 0x0A,
+    ClassCode = 0x0B,
+    CacheLineSize = 0x0C,
+    LatencyTimer = 0x0D,
+    HeaderType = 0x0E,
+    BIST = 0x0F,
+
+    // The next set of registers are for the 0x00 (standard) header.
+    // This currently uses only the common registers above that are available to all header types.
+
+    BaseAddr0 = 0x10,
+    BaseAddr1 = 0x14,
+    BaseAddr2 = 0x18,
+    BaseAddr3 = 0x1C,
+    BaseAddr4 = 0x20,
+    BaseAddr5 = 0x24,
+    CardbusCISPtr = 0x28,
+    SubsystemVenderId = 0x2C,
+    SubsystemId = 0x2E,
+    ExpansionROMBaseAddr = 0x30,
+    CapabilitiesPtr = 0x34,
+    InterruptLine = 0x3C,
+    InterruptPin = 0x3D,
+    MinGrant = 0x3E,
+    MaxLatency = 0x3F,
+
+    ///
+    /// Get the type the represents the width of the register. This can be either u8, u16 or u32.
+    ///
+    /// Argument:
+    ///     IN comptime pci_reg: PciRegisters - The register to get the width for.
+    ///
+    /// Return: type
+    ///     The width type.
+    ///
+    pub fn getWidth(comptime pci_reg: PciRegisters) type {
+        return switch (pci_reg) {
+            .RevisionId, .ProgrammingInterface, .Subclass, .ClassCode, .CacheLineSize, .LatencyTimer, .HeaderType, .BIST, .InterruptLine, .InterruptPin, .MinGrant, .MaxLatency, .CapabilitiesPtr => u8,
+            .VenderId, .DeviceId, .Command, .Status, .SubsystemVenderId, .SubsystemId => u16,
+            .BaseAddr0, .BaseAddr1, .BaseAddr2, .BaseAddr3, .BaseAddr4, .BaseAddr5, .CardbusCISPtr, .ExpansionROMBaseAddr => u32,
+        };
+    }
+};
+
+/// The PCI address used for sending to the address port.
+const PciAddress = packed struct {
+    register_offset: u8,
+    function: u3,
+    device: u5,
+    bus: u8,
+    reserved: u7 = 0,
+    enable: u1 = 1,
+};
+
+/// A PCI device. This will be unique to a bus and device number.
+const PciDevice = struct {
+    /// The bus on which the device is on
+    bus: u8,
+
+    /// The device number.
+    device: u5,
+
+    const Self = @This();
+
+    ///
+    /// Get the PCI address for this device and for a function and register.
+    ///
+    /// Argument:
+    ///     IN self: Self                     - This device.
+    ///     IN function: u3                   - The function.
+    ///     IN comptime pci_reg: PciRegisters - The register.
+    ///
+    /// Return: PciAddress
+    ///     The PCI address that can be used to read the register offset for this device and function.
+    ///
+    pub fn getAddress(self: Self, function: u3, comptime pci_reg: PciRegisters) PciAddress {
+        return PciAddress{
+            .bus = self.bus,
+            .device = self.device,
+            .function = function,
+            .register_offset = @enumToInt(pci_reg),
+        };
+    }
+
+    ///
+    /// Read the configuration register date from this device, function and register. This will
+    /// return the types width of the register, as PCI reads can only be done on a u32 basis.
```

Comment:
I don't quite understand the role of last part of this sentence. It says that reads can only be done with u32s, which I think is a bit confusing since it returns a integer of a type depending on the register width.

Reply:
Have improved the comment